components/modules/rmt.c

```diff
@@ -0,0 +1,403 @@
+
+  lua_pushvalue(L, 3);
+  params->cb_ref = luaL_ref(L, LUA_REGISTRYINDEX);
+  ud->rx_params = params;
+  params->channel = ud->channel;
+
+  xTaskCreate(handle_receive, "rmt-rx-receiver", 3000, params, 2, NULL);
+
+  return 0;
+}
+
+static void wait_for_transmit(void *param) {
+  lrmt_cb_params *p = (lrmt_cb_params *) param;
+  esp_err_t rc = rmt_wait_tx_done(p->channel, 10000 / portTICK_PERIOD_MS);
+
+  p->rc = rc;
+  task_post_high(cb_task_id, (task_param_t) p);
+
+  /* Destroy this task */
+  vTaskDelete(NULL);
+}
+
+static int lrmt_send(lua_State *L) {
+  lrmt_channel_t ud = (lrmt_channel_t)luaL_checkudata(L, 1, "rmt.channel");
+  if (!ud->tx) {
+    return luaL_error(L, "Cannot send on an RX channel");
+  }
+  int string_index = 2;
+
+  if (lua_type(L, 2) == LUA_TTABLE) {
+    lua_getfield(L, 2, "concat");
+    lua_pushvalue(L, 2);
+    lua_pushstring(L, "");
+    lua_call(L, 2, 1);
+    string_index = -1;
+  }
+
+  size_t len;
+  const char *data = lua_tolstring(L, string_index, &len);
+  if (!data || !len) {
+    return 0;
+  }
+
+  if (len & 1) {
+    return luaL_error(L, "Length must be a multiple of 2");
+  }
+
+  if (len & 3) {
+    // Just tack on a "\0\0"
+    luaL_Buffer b;
+    luaL_buffinit(L, &b);
+    luaL_addlstring(&b, data, len);
+    luaL_addlstring(&b, "\0\0", 2);
+    luaL_pushresult(&b);
+    data = lua_tolstring(L, -1, &len);
+    string_index = -1;
+  }
+
+  bool wait_for_done = true;
+
+  if (lua_type(L, 3) == LUA_TFUNCTION) {
+    // We have a callback
+    lrmt_cb_params *params = (lrmt_cb_params *) malloc(sizeof(*params));
+    if (!params) {
+      return luaL_error(L, "Cannot allocate memory");
+    }
+    memset(params, 0, sizeof(*params));
+
+    params->channel = ud->channel;
+
+    lua_pushvalue(L, 3);
+    params->cb_ref = luaL_ref(L, LUA_REGISTRYINDEX);
+    lua_pushvalue(L, string_index);
+    params->data_ref = luaL_ref(L, LUA_REGISTRYINDEX);
+    xTaskCreate(wait_for_transmit, "rmt-tx-waiter", 1024, params, 2, NULL);
+    wait_for_done = false;
+  }
+
+  // We want to transmit it
+  rmt_write_items(ud->channel, (rmt_item32_t *) data, len / sizeof(rmt_item32_t), wait_for_done);
+  return 0;
+}
+
+static int lrmt_close(lua_State *L) {
+  lrmt_channel_t ud = (lrmt_channel_t)luaL_checkudata(L, 1, "rmt.channel");
+
+  if (ud->channel >= 0) {
+    if (ud->rx_params) {
+      // We need to stop the listening task
+      ud->rx_params->rx_shutting_down = true;
+    } else {
+      rmt_driver_uninstall(ud->channel);
+      platform_rmt_release(ud->channel);
+    }
+    ud->channel = -1;
+  }
+
+  return 0;
+}
+
+static void cb_task(task_param_t param, task_prio_t prio) {
+  lrmt_cb_params *p = (lrmt_cb_params *) param;
+  lua_State *L = lua_getstate();
+
+  if (!p->dont_call) {
+    lua_rawgeti (L, LUA_REGISTRYINDEX, p->cb_ref);
+    if (p->data) {
+      lua_pushlstring(L, (char *) p->data, p->len);
+    } else {
+      lua_pushinteger(L, p->rc);
+    }
+
+    lua_call(L, 1, 0);
```

Comment:
You'll want `luaL_pcallx()` here to tap into the full error handling framework for callbacks.